### PR TITLE
Math gadgets refactory

### PR DESCRIPTION
**Pull request by smtmfft**
Addressing #586 and #900 
Make each gadget have its own file and own unit test, for better code organization & test coverage.
Meanwhile, keep the pub mods name consistent between before & after to avoid large scale code changes in evm_circuit.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+                let offset = 0;
+                let x = input_words[0];
+                let x_abs = input_words[1];
+                self.absword_gadget.assign(region, offset, x, x_abs)?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:131]: Can use some more tests with bigger numbers/negative numbers (try to also test edge cases in the implementation).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    },
+    util::Expr,
+};
+use eth_types::{Field, ToLittleEndian, ToScalar, Word};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+/// Construction of 2 256-bit words addition and result, which is useful for
+/// opcode ADD, SUB and balance operation
+#[derive(Clone, Debug)]
+pub(crate) struct AddWordsGadget<F, const N_ADDENDS: usize, const CHECK_OVREFLOW: bool> {
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:15]:
```suggestion
pub(crate) struct AddWordsGadget<F, const N_ADDENDS: usize, const CHECK_OVERFLOW: bool> {
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::MAX - 1, Word::from(1), Word::MAX],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr>>(
+            vec![Word::MAX, Word::from(1), Word::from(0)],
+            false,
+        );
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:230]: A test like this without `CHECK_OVREFLOW` enabled would interesting as well.

Also some tests with `N_ADDENDS` set to something `> 2` would be good.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    n.to_le_bytes().map(|byte| F::from(byte as u64)),
+                )?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, IsZeroGadgetTestContainer<Fr>>(vec![Word::from(0)], true);
+
+        test_math_gadget_container::<Fr, IsZeroGadgetTestContainer<Fr>>(vec![Word::from(1)], false);
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs:125]: Could use a test case where only a single value wasn't `0` or something like that.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/comparison.rs [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        test_math_gadget_container::<Fr, ComparisonTestContainer<Fr, false>>(
+            vec![Word::from(1), Word::from(0)],
+            false,
+        );
+
+        test_math_gadget_container::<Fr, ComparisonTestContainer<Fr, false>>(
+            vec![Word::from(10000), Word::from(2 << N)],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/comparison.rs:155]: Same as with `CmpWordsGadget`.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 5, 4>>(
+            vec![Word::from(1)],
+            false,
+        );
+
+        test_math_gadget_container::<Fr, ConstantDivisionTestContainer<Fr, 16, 0>>(
+            vec![Word::from(1 << N)],
+            true,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs:162]: I think could use some extra test cases that checks that
- remainder < denominator
- Require that quotient < 2**N_BYTES
- The quotient indeed is incorrect (guess this one may be a trickier because needs changing the generated witness)

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        }
+
+        impl<F: Field> MathGadgetContainer<F> for IsEqualGadgetTestContainer<F> {
+            const NAME: &'static str = "IsEqualGadget";
+
+            fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+                let a = cb.query_rlc();
+                let b = cb.query_rlc();
+                let eq_gadget =
+                    IsEqualGadget::<F>::construct(cb, sum::expr(&a.cells), sum::expr(&b.cells));
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs:64]: This looks a bit strange, the gadget works on simple field elements so wouldn't expect words to be used here.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            ) -> Result<(), Error> {
+                let a = input_words[0];
+                let b = input_words[1];
+                let offset = 0;
+
+                self.a.assign(region, offset, Some(a.to_le_bytes()))?;
+                self.b.assign(region, offset, Some(b.to_le_bytes()))?;
+                self.eq_gadget.assign(
+                    region,
+                    0,
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs:82]:
```suggestion
                    region,
                    offset,
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            z_gadget: IsZeroGadget<F>,
+            n: util::Word<F>,
+        }
+
+        impl<F: Field> MathGadgetContainer<F> for IsZeroGadgetTestContainer<F> {
+            const NAME: &'static str = "IsZeroGadget";
+
+            fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+                let n = cb.query_word();
+                let z_gadget = IsZeroGadget::<F>::construct(cb, sum::expr(&n.cells));
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs:82]: Same as with `IsEqualGadget`, just use a single cell/field element.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+                self.n.assign(region, offset, Some(n.to_le_bytes()))?;
+                self.z_gadget
+                    .assign(region, 0, sum::value(&n.to_le_bytes()))?;
+
+                Ok(())
+            }
+        }
+
+        test_math_gadget_container::<Fr, IsZeroGadgetTestContainer<Fr>>(vec![Word::from(0)], false);
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs:103]: Strange input values where it's the opposite of what you'd expect (0 -> true).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/less.rs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+            false,
+        );
+
+        // out of range check
+        test_math_gadget_container::<Fr, LtGadgetTestContainer<Fr>>(
+            vec![Word::from(1), Word::from(2 << (N * 8))],
+            false,
+        );
+    }
+}
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/less.rs:159]: Would call this file `less_than.rs` or `lt.rs` because it's the more standard way to call this (unless there's a reason not to e.g. a conflict or something).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/less_word.rs [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        test_math_gadget_container::<Fr, LtWordTestContainer<Fr>>(
+            vec![Word::from(1), Word::from(0)],
+            false,
+        );
+
+        test_math_gadget_container::<Fr, LtWordTestContainer<Fr>>(
+            vec![Word::MAX, Word::MAX],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/less_word.rs:138]: Could also benefit of some `LO_MAX`/`HI_MAX` like testing I think.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+
+                self.a.assign(region, offset, Value::known(a))?;
+                self.b.assign(region, offset, Value::known(b))?;
+                self.minmax_gadget.assign(region, offset, a, b)?;
+
+                Ok(())
+            }
+        }
+
+        // min == a
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs:112]:
```suggestion
        // min == a, max == b
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            vec![Word::from(1), Word::from(0)],
+            false,
+        );
+
+        // min == b
+        test_math_gadget_container::<Fr, MinMaxTestContainer<Fr, false>>(
+            vec![Word::from(1), Word::from(0)],
+            true,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs:133]: Although not directly interesting, just some non-binary test values would be great.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mod_gadget.rs [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+        test_math_gadget_container::<Fr, ModGadgetTestContainer<Fr>>(
+            vec![Word::from(1), Word::from(1), Word::from(0)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, ModGadgetTestContainer<Fr>>(
+            vec![Word::from(1), Word::from(1), Word::from(1)],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/mod_gadget.rs:167]: Please add some higher value test cases.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+        test_math_gadget_container::<Fr, MulAddGadgetContainer<Fr>>(
+            vec![Word::from(1), Word::from(1), Word::from(1), Word::from(2)],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, MulAddGadgetContainer<Fr>>(
+            vec![Word::from(10), Word::from(1), Word::from(1), Word::from(3)],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs:232]: Please add some higher value test cases (also LO_MAX/HI_MAX values seem interesting here).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words512.rs [hunk 17]
```diff
@@ -0,0 +1,317 @@
...
+            vec![
+                Word::from(10),
+                Word::from(1),
+                Word::from(1),
+                Word::from(3),
+                Word::from(0),
+            ],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words512.rs:316]: Please add some higher value test cases (also LO_MAX/HI_MAX values seem interesting here).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mul_word_u64.rs [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+        test_math_gadget_container::<Fr, MulWordByU64TestContainer<Fr>>(
+            vec![Word::MAX, Word::from(1), Word::MAX],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, MulWordByU64TestContainer<Fr>>(
+            vec![Word::MAX, Word::from(1), Word::from(1)],
+            false,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/mul_word_u64.rs:173]: Please add some higher value test cases (also LO_MAX/HI_MAX values seem interesting here).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/pair_select.rs [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+            vec![Word::from(0), Word::from(1), Word::from(0)],
+            false,
+        );
+
+        // choose b check
+        test_math_gadget_container::<Fr, PairSelectionTestContainer<Fr, false>>(
+            vec![Word::from(0), Word::from(1), Word::from(0)],
+            true,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/pair_select.rs:151]: Some non-binary test values always a good idea.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs (hunk 1 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:131]: Adding minus tests

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs (hunk 2 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:15]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs (hunk 4 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs:125]: 1 is that case, but it a little bit special as the bit pos is in the first pos. will add a number like 1024 for same purpose test.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs (hunk 3 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:230]: Sure, will add more

File: zkevm-circuits/src/evm_circuit/util/math_gadget/cmp_words.rs [hunk 20]
```diff
@@ -0,0 +1,181 @@
...
+        test_math_gadget_container::<Fr, CmpWordGadgetTestContainer<Fr, false>>(
+            vec![Word::from(1), Word::MAX],
+            true,
+        );
+
+        test_math_gadget_container::<Fr, CmpWordGadgetTestContainer<Fr, false>>(
+            vec![Word::from(1), Word::from(0)],
+            false,
+        );
+    }
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/cmp_words.rs:180]: Added.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs (hunk 7 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs:64]: Changed to cell instead of RLC.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs (hunk 8 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs:82]: fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs (hunk 9 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs:82]: fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs (hunk 10 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/is_zero.rs:103]: Changed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs (hunk 6 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/constant_division.rs:162]: add more cases for these conditions.

**Comment by smtmfft** [comment]
All issues are addressed, and re-org the test code for better self explanation.

File: zkevm-circuits/src/evm_circuit/util/math_gadget.rs [hunk 21]
```diff
@@ -1,753 +1,49 @@
...
+mod byte_size;
+mod cmp_words;
+mod comparison;
+mod constant_division;
+mod is_equal;
+mod is_zero;
+mod lt;
+mod lt_word;
+mod min_max;
+mod mod_gadget;
```
Comment (johntaiko) [inline zkevm-circuits/src/evm_circuit/util/math_gadget.rs:21]:
```suggestion
mod modulo;
```
Naming style: `mod_gadget` vs `modulo`?(anyway `mod` is a keyword of rust

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        );
+    }
+
+    #[test]
+    fn test_abs_wordmax_eq_minus1() {
+        test_math_gadget_container::<Fr, AbsWordGadgetContainer<Fr>>(
+            vec![Word::MAX, Word::from(1)],
+            true,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:159]: Hmmm some `MAX_HI`/`MAX_LO` testing would be great here as well!
Another thing that I see is that `is_neg(&self)` is actually used externally as well so the test ideally also checks that this flag is set correctly.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs [hunk 23]
```diff
@@ -0,0 +1,294 @@
...
+
+            let addends = input_words[0..N_ADDENDS].try_into().unwrap();
+            self.addwords_gadget.assign(region, 0, addends, sum)?;
+
+            Ok(())
+        }
+    }
+
+    #[test]
+    fn test_add_0_0() {
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:207]:
```suggestion
    fn test_addwords_0_0() {
```

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs [hunk 24]
```diff
@@ -0,0 +1,294 @@
...
+        );
+    }
+
+    #[test]
+    fn test_addwords_3_addends_without_overflow_check() {
+        test_math_gadget_container::<Fr, AddWordsTestContainer<Fr, 3, false>>(
+            vec![Word::MAX, Word::from(1), Word::from(0), Word::from(0)],
+            true,
+        );
+    }
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs:293]: `MAX_HI`/`MAX_LO` testing would be great here as well.
Also `carry(&self)` at least (maybe some of the others as well) are used externally so ideally tested here.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs (hunk 4 above)
Comment (Brechtpd) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs:125]: Oh I see, I didn't notice that the byte values of the word were passed into the gadget. Is there a reason to do it like that? It's a bit confusing and doesn't seem to make testing easier (harder to specify different values).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs [hunk 25]
```diff
@@ -0,0 +1,162 @@
...
+            true,
+        );
+        test_math_gadget_container::<Fr, MinMaxTestContainer<Fr, true>>(
+            vec![Word::from(3), Word::from(5)],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_minmx_unexpect_min_a() {
```
Comment (Brechtpd) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/min_max.rs:137]:
```suggestion
    fn test_minmax_unexpect_min_a() {
```

**Comment by Brechtpd** [comment]
I think it would increase readability if all these `test_math_gadget_container::<Fr, _<Fr>>` wrappers, which are almost always the same in each test file, is wrapped in a short function/macro (like `try_test!` or something like that) so the useful data in each test is easier to parse and also less data duplicated.

The descriptive test names are a big improvement to easier see what's being tested. I think sometimes a comment would be even easier but I think like it is now is good enough.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs (hunk 22 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:159]: Added

File: zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs (hunk 24 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/addwords.rs]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs (hunk 4 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/batched_is_zero.rs:124]: That's because the WORD (U256) is by design to be a unified input param for all gadgets, as it can be easily changed to any other value type. The Pro is we have an all in one test framework, however, the Cons is a little bit confusing for some cases which do not take WORD. It well fits that old all-gadgets-in-one-file math_gadget.rs. Now I think it still fine, but yes, it can be managed better as each file has own code space.

**Review by Brechtpd — APPROVED** [review]

**Comment by Brechtpd** [comment]
One issue found and corrected:

https://github.com/privacy-scaling-explorations/zkevm-specs/blob/33e78c0d5f3baae05d6a51cedc6df387efa48532/specs/opcode/02MUL_04DIV_06MOD.md?plain=1#L60:
```
overflow = carry_{hi} + A_1B_3 + A_2B_2 + A_3B_1 + A_2B_3 + A_3B_2 + A_3B_3
```

vs

https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/e7560be5eeea63fc64bd3a3d45ba24784e67f122/zkevm-circuits/src/evm_circuit/util/math_gadget.rs#L840-L846

Line 843 should be `+ a_limbs[3].clone() * b_limbs[1].clone() `.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs [hunk 26]
```diff
@@ -0,0 +1,302 @@
...
+    }
+
+    #[test]
+    fn test_addwords_1000_1000() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true>, vec![Word::from(1000), Word::from(1000), Word::from(2000)], true);
+    }
+
+    #[test]
+    fn test_addwords_to_wordmax() {
+        try_test!(AddWordsTestContainer<Fr, 2,0u64,true,>, vec![Word::MAX - 1, Word::from(1), Word::MAX],true);
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs:234]: Nitpick: space missing before `true` (multiple occurrences also down below).

File: zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs [hunk 27]
```diff
@@ -0,0 +1,302 @@
...
+    fn test_addwords_wrong_sum0() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, [Word::from(1), Word::from(0),  Word::from(0)],false);
+    }
+
+    #[test]
+    fn test_addwords_wrong_sum2() {
+        try_test!(AddWordsTestContainer<Fr, 2, 0u64, true,>, [Word::from(2), Word::from(1),  Word::from(2)],false);
+    }
+
+    //non overflow check cases
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs:257]: Seems like 'overflow check cases'?

File: zkevm-circuits/src/evm_circuit/util/math_gadget/byte_size.rs [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+        bytesize_gadget: ByteSizeGadget<F>,
+        a: util::Word<F>,
+    }
+
+    impl<F: Field, const N: u8> MathGadgetContainer<F> for ByteSizeGadgetContainer<F, N> {
+        fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+            let value_rlc = cb.query_word();
+            let bytesize_gadget = ByteSizeGadget::<F>::construct(cb, &value_rlc);
+            cb.require_equal(
+                "byte size gadget must equal",
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/byte_size.rs:129]: 'must equal N'

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs [hunk 29]
```diff
@@ -0,0 +1,174 @@
...
+        try_test!(AbsWordGadgetContainer<Fr, true>, [WORD_HIGH_MAX, abs_high_max], true);
+    }
+
+    #[test]
+    fn test_abs_word_low_max() {
+        try_test!(AbsWordGadgetContainer<Fr, false>, [WORD_LOW_MAX, WORD_LOW_MAX], true);
+    }
+
+    #[test]
+    fn test_abs_unexpected_is_neg_for_postive() {
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:166]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs [hunk 30]
```diff
@@ -0,0 +1,174 @@
...
+        try_test!(AbsWordGadgetContainer<Fr, true>, [WORD_LOW_MAX, WORD_LOW_MAX], false);
+    }
+
+    #[test]
+    fn test_abs_unexpected_is_neg_for_negtive() {
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:171]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs (hunk 27 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs:257]: This comment is outdated as cases growing, deleted.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/byte_size.rs (hunk 28 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/byte_size.rs:129]: Fixed.

File: zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs (hunk 26 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/add_words.rs]: Fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs [hunk 31]
```diff
@@ -0,0 +1,127 @@
...
+        a: Cell<F>,
+        b: Cell<F>,
+    }
+
+    impl<F: Field> MathGadgetContainer<F> for IsEqualGadgetTestContainer<F> {
+        fn configure_gadget_container(cb: &mut ConstraintBuilder<F>) -> Self {
+            let a = cb.query_cell();
+            let b = cb.query_cell();
+            let eq_gadget = IsEqualGadget::<F>::construct(cb, a.expr(), b.expr());
+            cb.require_equal("Inputs must equal", eq_gadget.expr(), 1.expr());
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/is_equal.rs:62]: Nitpick: a bit unclear message, perhaps "Inputs must be equal (a == b)"?

File: zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs [hunk 32]
```diff
@@ -0,0 +1,212 @@
+use crate::{
+    evm_circuit::util::{
+        self, constraint_builder::ConstraintBuilder, math_gadget::*, sum, CachedRegion,
+    },
+    util::Expr,
+};
+use eth_types::{Field, ToLittleEndian, Word};
+use halo2_proofs::plonk::Error;
+
+/// Constraints the words a, n, r such that:
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs:10]: "Constrain the words a, n, r such that:" or "Constraints for the words a, n, r:"

File: zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs [hunk 33]
```diff
@@ -0,0 +1,212 @@
...
+/// a mod n = r, if n!=0
+/// r = 0,       if n==0
+///
+/// We use the auxiliary a_or_zero word, whose value is constrained to be: a if
+/// n!=0, 0 if n==0. This allows to use the equation k * n + r = a_or_zero to
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs:15]: Perhaps use spaces consistently (between variables and operators)

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs [hunk 34]
```diff
@@ -0,0 +1,376 @@
...
+            vec![
+                Word::from(100),
+                Word::from(54),
+                WORD_LOW_MAX,
+                Word::from(5400) + WORD_LOW_MAX,
+                Word::from(0)
+            ],
+            true,
+        );
+        // 100 * 54 + low_max == low_max + 5400
```
Comment (miha-stopar) [inline zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs:286]: high_max

**Review by miha-stopar — APPROVED** [review]
Just nitpicks, otherwise LGTM!

File: zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs (hunk 32 above)
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/modulo.rs:10]: fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs [hunk 35]
```diff
@@ -0,0 +1,376 @@
...
+                Word::from(10),
+                Word::from(1),
+                Word::from(1),
+                Word::from(3),
+                Word::from(0)
+            ],
+            false,
+        );
+
+        // 100 * 54 + low_max == low_max + 5400, no overflow
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/mul_add_words.rs:350]: fixed

File: zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+#[derive(Clone, Debug)]
+pub(crate) struct AbsWordGadget<F> {
+    x: util::Word<F>,
+    x_abs: util::Word<F>,
+    sum: util::Word<F>,
+    is_neg: LtGadget<F, 1>,
+    add_words: AddWordsGadget<F, 2, false>,
+}
+
+impl<F: Field> AbsWordGadget<F> {
```
[the remark this answers is not in the document]
Comment (smtmfft) [reply zkevm-circuits/src/evm_circuit/util/math_gadget/abs_word.rs:20]: OK, let just keep the status this time :)

**Comment by ChihChengLiang** [comment]
@smtmfft Looks like mergeable with 1 final rebase 🎉 .
